### PR TITLE
INHEART-138: Add Passport.js to API

### PR DESCRIPTION
**Pull request by cierzniak**
## Background

As software engineers we want to guard our application and disallow everyone to access to it. We can achieve it by adding authorization layer. To make it more secure we can use external CAS (Central Authentication Service). Using internal company CAS helps us also keep up-to-date who's an employee and employees aren't forced to create next account with one more password.

## In this PR

Th PR contains new module to backend - AuthModule.

AuthModule connects to internal OAuth2 provider (Keycloak) which is available as Docker image. I used [official NestJS documentation](https://docs.nestjs.com/security/authentication) and [sdoxsee's article from 2020](https://sdoxsee.github.io/blog/2020/02/05/cats-nest-nestjs-mongo-oidc.html).

After fetching user info from OAuth2 new JWT is generated. Via JWT we can use rest of our endpoints.

Keycloak have initial realm configuration stored in `/docker/keycloak`. By default it's configured to provide two users:
* `admin@inheartive.io` : `Password1`
* `user@inheartive.io` : `Password1`

**Keycloak is available at http://localhost:8080.**

### CODE REVIEW

File: apps/backend/src/auth/model/oidc.token.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { UserinfoResponse } from 'openid-client';
+
+export interface OidcToken {
```
Comment (cierzniak) [inline apps/backend/src/auth/model/oidc.token.ts:3]: I'm not sure this file is correctly named (`/model/oidc.token`).

**Review by KonradKarlinski — APPROVED** [review]

File: apps/backend/src/auth/login.guard.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import { ExecutionContext, Injectable } from '@nestjs/common';
+import { AuthGuard } from '@nestjs/passport';
+
+@Injectable()
+export class LoginGuard extends AuthGuard('oidc') {
+  async canActivate(context: ExecutionContext) {
+    const result = (await super.canActivate(context)) as boolean;
+    const request = context.switchToHttp().getRequest();
```
[the remark this answers is not in the document]
Comment (pmusialik-intive) [reply apps/backend/src/auth/login.guard.ts:8]: nice

**Review by zgr92 — COMMENTED** [review]
I don't know Nest.js, but code looks good to me, it's clear, Readme is updated and the logic make sense to me.

**Review by zgr92 — APPROVED** [review]

File: apps/backend/src/auth/login.guard.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (cierzniak) [reply apps/backend/src/auth/login.guard.ts:8]: Ok, I know why we change context to HTTP. In [documentation](https://docs.nestjs.com/fundamentals/execution-context#current-application-context) I've found this and it's connected to multiple communication ways in NestJS (REST/gRCP/GraphQL).

**Comment by konradhanus** [comment]
- [ ] - add tests

**Review by pmusialik-intive — APPROVED** [review]

**Pull request merged**